ext/ffi/lib.rs

```diff
@@ -535,322 +564,792 @@ where
+    },
+    NativeType::U32 => NativeValue {
+      u32_value: unsafe { cif.call::<u32>(fun_ptr, &call_args) },
+    },
+    NativeType::I32 => NativeValue {
+      i32_value: unsafe { cif.call::<i32>(fun_ptr, &call_args) },
+    },
+    NativeType::U64 => NativeValue {
+      u64_value: unsafe { cif.call::<u64>(fun_ptr, &call_args) },
+    },
+    NativeType::I64 => NativeValue {
+      i64_value: unsafe { cif.call::<i64>(fun_ptr, &call_args) },
+    },
+    NativeType::USize => NativeValue {
+      usize_value: unsafe { cif.call::<usize>(fun_ptr, &call_args) },
+    },
+    NativeType::ISize => NativeValue {
+      isize_value: unsafe { cif.call::<isize>(fun_ptr, &call_args) },
+    },
+    NativeType::F32 => NativeValue {
+      f32_value: unsafe { cif.call::<f32>(fun_ptr, &call_args) },
+    },
+    NativeType::F64 => NativeValue {
+      f64_value: unsafe { cif.call::<f64>(fun_ptr, &call_args) },
+    },
+    NativeType::Pointer | NativeType::Function {} => NativeValue {
+      pointer: unsafe { cif.call::<*const u8>(fun_ptr, &call_args) },
+    },
+  })
+}
 
-  for (&native_type, value) in symbol
-    .parameter_types
-    .iter()
-    .zip(args.parameters.into_iter())
-  {
-    match native_type {
-      NativeType::Pointer => match value.as_u64() {
-        Some(idx) => {
-          let buf = buffers
-            .get(idx as usize)
-            .ok_or_else(|| {
-              generic_error(format!("No buffer present at index {}", idx))
-            })?
-            .unwrap();
-          native_values.push(NativeValue::buffer(buf.as_ptr()));
-        }
-        _ => {
-          let value = NativeValue::new(native_type, value)?;
-          native_values.push(value);
-        }
-      },
+struct RegisteredCallbackResource {
+  closure: libffi::middle::Closure<'static>,
+  info: *const CallbackInfo,
+}
+
+impl Resource for RegisteredCallbackResource {
+  fn name(&self) -> Cow<str> {
+    "registeredcallback".into()
+  }
+
+  fn close(self: Rc<Self>) {
+    let info = unsafe { Box::from_raw(self.info as *mut CallbackInfo) };
+    let isolate = unsafe { info.isolate.as_mut().unwrap() };
+    unsafe { v8::Global::from_raw(isolate, info.callback) };
+    unsafe { v8::Global::from_raw(isolate, info.context) };
+    drop(self)
+  }
+}
+
+struct CallbackInfo {
+  pub callback: NonNull<v8::Function>,
+  pub context: NonNull<v8::Context>,
+  pub isolate: *mut v8::Isolate,
+}
+
+unsafe extern "C" fn deno_ffi_callback(
+  cif: &libffi::low::ffi_cif,
+  result: &mut c_void,
+  args: *const *const c_void,
+  info: &CallbackInfo,
+) {
+  let isolate = &mut *info.isolate;
+  let callback = v8::Global::from_raw(isolate, info.callback);
+  let context = std::mem::transmute::<
+    NonNull<v8::Context>,
+    v8::Local<v8::Context>,
+  >(info.context);
+  IS_EVENT_LOOP_THREAD.with(|is_event_loop_thread| {
+    if !(*is_event_loop_thread.borrow()) {
+      // Call from another thread, PANIC IN THE DISCO!
+      todo!("Call from another thread");
```

Comment:
UB to unwind panic from across FFI. Maybe we can change this to an exit(1)

Reply:
I guess an `eprintln!` or something like that would be useful before exiting?

Reply:
sure